flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinatorTriggeringTest.java

```diff
@@ -283,31 +300,494 @@ public void testStopPeriodicScheduler() throws Exception {
+			assertTrue(checkpointExceptionOptional.isPresent());
+			assertEquals(CheckpointFailureReason.PERIODIC_SCHEDULER_SHUTDOWN,
+				checkpointExceptionOptional.get().getCheckpointFailureReason());
 		}
 
 		// Not periodic
+		final CompletableFuture<CompletedCheckpoint> onCompletionPromise2 = coord.triggerCheckpoint(
+			System.currentTimeMillis(),
+			CheckpointProperties.forCheckpoint(CheckpointRetentionPolicy.NEVER_RETAIN_AFTER_TERMINATION),
+			null,
+			false,
+			false);
+		manuallyTriggeredScheduledExecutor.triggerAll();
+		assertFalse(onCompletionPromise2.isCompletedExceptionally());
+	}
+
+	@Test
+	public void testTriggerCheckpointWithShuttingDownCoordinator() throws Exception {
+		// create some mock Execution vertices that receive the checkpoint trigger messages
+		final ExecutionAttemptID attemptID1 = new ExecutionAttemptID();
+		ExecutionVertex vertex1 = mockExecutionVertex(attemptID1);
+
+		// set up the coordinator and validate the initial state
+		CheckpointCoordinatorConfiguration chkConfig = new CheckpointCoordinatorConfiguration(
+			600000,
+			600000,
+			0,
+			Integer.MAX_VALUE,
+			CheckpointRetentionPolicy.NEVER_RETAIN_AFTER_TERMINATION,
+			true,
+			false,
+			0);
+		CheckpointCoordinator coord = new CheckpointCoordinator(
+			new JobID(),
+			chkConfig,
+			new ExecutionVertex[] { vertex1 },
+			new ExecutionVertex[] { vertex1 },
+			new ExecutionVertex[] { vertex1 },
+			new StandaloneCheckpointIDCounter(),
+			new StandaloneCompletedCheckpointStore(1),
+			new MemoryStateBackend(),
+			Executors.directExecutor(),
+			manuallyTriggeredScheduledExecutor,
+			SharedStateRegistry.DEFAULT_FACTORY,
+			failureManager);
+
+		coord.startCheckpointScheduler();
+		// Periodic
+		final CompletableFuture<CompletedCheckpoint> onCompletionPromise = coord.triggerCheckpoint(
+			System.currentTimeMillis(),
+			CheckpointProperties.forCheckpoint(CheckpointRetentionPolicy.NEVER_RETAIN_AFTER_TERMINATION),
+			null,
+			true,
+			false);
+
+		coord.shutdown(JobStatus.FAILED);
+		manuallyTriggeredScheduledExecutor.triggerAll();
 		try {
-			coord.triggerCheckpoint(
-				System.currentTimeMillis(),
-				CheckpointProperties.forCheckpoint(CheckpointRetentionPolicy.NEVER_RETAIN_AFTER_TERMINATION),
-				null,
-				false,
-				false);
-			manuallyTriggeredScheduledExecutor.triggerAll();
-		} catch (CheckpointException e) {
-			fail("Unexpected exception : " + e.getCheckpointFailureReason().message());
+			onCompletionPromise.get();
+			fail("Should not reach here");
+		} catch (ExecutionException e) {
+			final Optional<CheckpointException> checkpointExceptionOptional =
+				ExceptionUtils.findThrowable(e, CheckpointException.class);
+			assertTrue(checkpointExceptionOptional.isPresent());
+			assertEquals(CheckpointFailureReason.CHECKPOINT_COORDINATOR_SHUTDOWN,
+				checkpointExceptionOptional.get().getCheckpointFailureReason());
 		}
 	}
 
+	@Test
+	public void testTriggerCheckpointRequestQueued() throws Exception {
```

Comment:
optional nit: rename `testTriggerCheckpointRequestQueued` -> `testTriggerCheckpointBeforePreviousOneCompleted`? (to name it after a behaviour/test case instead of a private field name?)